### PR TITLE
Add left and right buttons to slides component

### LINKED ISSUE #18: Home page flip through event posters
image

### PR DESCRIPTION
**Pull request by zakpatterson**
Fixes #18

### CODE REVIEW

File: packages/site/src/lib/components/Slides.svelte [hunk 1]
```diff
@@ -1,35 +1,66 @@
...
-      altText: string;
-      imageUrl: string;
-    };
+    poster?: Poster
+  }
+
+  const posters = (): Poster[] => events.flatMap((ev) => ev.poster ? [ev.poster] : [])
+
+  let currentIdx = 0;
+  let current: Poster | undefined = posters()[0]
```
Comment (samwightt) [inline packages/site/src/lib/components/Slides.svelte]:
```suggestion
  let currentIdx = 0;
  $: current: Poster | undefined = currentIdx >= 0 && currentIdx < posters.length ? posters[currentIdx] : undefined
```

[Svelte keeps derived data like this up-to-date with its reactive declaration syntax](https://svelte.dev/tutorial/reactive-declarations). Anytime `currentIdx` is updated, `current` will be updated automatically.

Also moved the bounds check into here just to make sure `currentIdx` is in bounds (if the events array has zero elements then `current` should be `undefined`).

File: packages/site/src/lib/components/Slides.svelte [hunk 2]
```diff
@@ -1,35 +1,66 @@
...
 
   interface Event {
-    poster?: {
-      altText: string;
-      imageUrl: string;
-    };
+    poster?: Poster
+  }
+
+  const posters = (): Poster[] => events.flatMap((ev) => ev.poster ? [ev.poster] : [])
```
Comment (samwightt) [inline packages/site/src/lib/components/Slides.svelte]:
```suggestion
  $: posters: Poster[] = events.flatMap((ev) => ev.poster ? [ev.poster] : [])
```

This should use the [reactive declaration syntax](https://svelte.dev/tutorial/reactive-declarations) so Svelte knows to update it when the props update (props are reactive and Svelte might share a component instance and just swap props).

Also TIL flatMap can be used as a filtermap!!

File: packages/site/src/lib/components/Slides.svelte [hunk 3]
```diff
@@ -1,35 +1,66 @@
...
+  let currentIdx = 0;
+  let current: Poster | undefined = posters()[0]
+
+  const advance = (): void => {
+    const ps = posters()
+    if (currentIdx < ps.length - 1) {
+      currentIdx++;
+      current = ps[currentIdx];
+    }
+  }
```
Comment (samwightt) [inline packages/site/src/lib/components/Slides.svelte]:
```suggestion
  const advance = (): void => {
    if (canAdvance) {
      currentIdx++;
    }
  }
```

Svelte creates these functions in a way such that the reference to `posters` and `currentPoster` will always be up-to-date. `current` will also be automatically re-run.

File: packages/site/src/lib/components/Slides.svelte [hunk 4]
```diff
@@ -1,35 +1,66 @@
...
+
+  const goBack = (): void => { 
+    const ps = posters()
+    if(currentIdx > 0) {
+      currentIdx--;
+      current = ps[currentIdx];
+    }
   }
```
Comment (samwightt) [inline packages/site/src/lib/components/Slides.svelte]:
```suggestion
  const goBack = (): void => {
    if (canGoBack) {
      currentIdx--;
    }
  }
```

File: packages/site/src/lib/components/Slides.svelte [hunk 5]
```diff
@@ -1,35 +1,66 @@
...
 
-  let currentEvent = events.length > 0 && events[0];
 </script>
 
-{#if currentEvent}
+{#if current}
   <div class="card">
-    {#if currentEvent.poster}
+    {#if current}
```
Comment (samwightt) [inline packages/site/src/lib/components/Slides.svelte]: This `if` is unnecessary.

File: packages/site/src/lib/components/Slides.svelte [hunk 6]
```diff
@@ -1,35 +1,66 @@
...
-      imageUrl: string;
-    };
+    poster?: Poster
+  }
+
+  const posters = (): Poster[] => events.flatMap((ev) => ev.poster ? [ev.poster] : [])
+
+  let currentIdx = 0;
+  let current: Poster | undefined = posters()[0]
+
```
Comment (samwightt) [inline packages/site/src/lib/components/Slides.svelte:21]:
```suggestion
$: canAdvance = currentIdx < posters.length - 1
$: canGoBack = currentIdx > 0
```

You have two unnamed concepts here, plus you write these twice.

File: packages/site/src/lib/components/Slides.svelte [hunk 7]
```diff
@@ -1,35 +1,66 @@
...
 
-  let currentEvent = events.length > 0 && events[0];
 </script>
 
-{#if currentEvent}
+{#if current}
   <div class="card">
-    {#if currentEvent.poster}
+    {#if current}
+      {#if currentIdx > 0}
```
Comment (samwightt) [inline packages/site/src/lib/components/Slides.svelte]:
```suggestion
      {#if canGoBack}
```

File: packages/site/src/lib/components/Slides.svelte [hunk 8]
```diff
@@ -1,35 +1,66 @@
...
       <Image
-        src={currentEvent.poster.imageUrl}
-        alt={currentEvent.poster.altText}
+        src={current.imageUrl}
+        alt={current.altText}
         width={600}
         quality={1}
         style="width: 100%; height: 100%; object-fit: cover; position: absolute; top: 0; left: 0; z-index: 1; filter: blur(10px);"
       />
+      {#if currentIdx < posters().length - 1 }
```
Comment (samwightt) [inline packages/site/src/lib/components/Slides.svelte]:
```suggestion
      {#if canAdvance}
```

File: packages/site/src/routes/+page.svelte [hunk 9]
```diff
@@ -24,7 +24,7 @@
   <CallToAction />
   <div class="sideby">
     {#if nextEvent}
```
Comment (samwightt) [inline packages/site/src/routes/+page.svelte:26]:
```suggestion
    {#if upcomingEvents.length > 0}
```

You can delete `nextEvent`!

**Review by samwightt — CHANGES_REQUESTED** [review]
This looks good! Sorry if this is a bit nit-picky, it's a habit cause the older devs at my work are nit-picky. Key thing here is just to move the derived data (`current` and `posters`) into reactive declarations so that Svelte can ensure they have the most up-to-date data. Thankfully in Svelte you don't have to worry about references like you have to in React, so `posters` being a function isn't needed (the reference to `events` in a reactive declaration is guaranteed to always be up-to-date).

File: packages/site/src/lib/components/Slides.svelte (hunk 1 above)
Comment (zakpatterson) [reply packages/site/src/lib/components/Slides.svelte]: Ok, thanks for explaining that to me, had no idea what the $: syntax meant at first

File: packages/site/src/lib/components/Slides.svelte (hunk 2 above)
Comment (zakpatterson) [reply packages/site/src/lib/components/Slides.svelte]: ha yea never `.filter().map()`

**Pull request merged**